core/src/main/java/net/opentsdb/query/processor/downsample/DownsampleNumericToNumericArrayIterator.java

```diff
@@ -0,0 +1,577 @@
+    } else {
+      iterator = null;
+    }
+
+    has_next = iterator != null ? iterator.hasNext() : false;
+
+  }
+
+  @Override
+  public boolean hasNext() {
+    return has_next;
+  }
+
+  /**
+   * The pre computed result in array aggregator
+   * 
+   * @param aggregator
+   * @return
+   */
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  @Override
+  public TimeSeriesValue<NumericArrayType> nextPool(Aggregator aggregator) {
+
+    has_next = false;
+
+    NumericArrayAggregator numericArrayAggregator = (NumericArrayAggregator) aggregator; // group by
+                                                                                         // aggregator
+    Accumulator accumulator = DownsampleNumericToNumericArrayIterator.threadLocalAccs.get();
+    accumulator.reset();
+
+    double[] localDoubleAggs = DownsampleNumericToNumericArrayIterator.threadLocalDoubleAggs.get();
+    long[] localLongAggs = DownsampleNumericToNumericArrayIterator.threadLocalLongAggs.get();
+
+    resetLocalAggs(localDoubleAggs, localLongAggs);
+
+    TimeStamp nextTs = null;
+
+    boolean flush = false;
+    int index = 0;
+    while (iterator.hasNext()) {
+      TimeSeriesValue<NumericType> next = (TimeSeriesValue<NumericType>) iterator.next();
+
+      if (next.timestamp().compare(Op.GTE, this.interval_start)
+          && next.timestamp().compare(Op.LT, this.interval_end)) {
+
+        if (nextTs == null) {
+          // Use the first valid dp timestamp as the first Timestamp.
+          nextTs = next.timestamp().getCopy();
+          long nextRoundoffTs =
+              next.timestamp().epoch() - (next.timestamp().epoch() % intervalPart) + intervalPart;
+          nextTs.updateEpoch(nextRoundoffTs);
+        }
+
+        if (next.timestamp().compare(Op.GTE, nextTs)) {
+          // flush previous interval once new interval starts
+          flush = true;
+
+          // find the index to update in the Aggregator
+          index = (int) (((nextTs.epoch() - interval_start.epoch() - intervalPart) / intervalPart)
+              % intervals);
+
+          // get the next timestamp in the downsample
+          long nextTsRoundoff =
+              next.timestamp().epoch() - (next.timestamp().epoch() % intervalPart) + intervalPart;
+          nextTs.updateEpoch(nextTsRoundoff);
+        }
+        
+        if (accumulator.isFull() || flush) {
+          accumulateDoubles(accumulator, localLongAggs, localDoubleAggs);
+          accumulateLongs(accumulator, localLongAggs, localDoubleAggs);
+
+          accumulator.reset();
+        }
+        if (next.value().isInteger()) {
+          accumulator.add(next.value().longValue());
+        } else {
+          accumulator.add(next.value().doubleValue());
+        }
+
+        if (flush) {
+
+          // increment interval in the groupby aggregator
+          double v = getAggValue(localDoubleAggs, localLongAggs);
+
+          if (numericArrayAggregator instanceof ArrayCount && this.aggEnum == AggEnum.count) {
```

Comment:
Add a TODO about making this generic. I don't like having specific aggregation callouts if at all possible. (even though we handle the sum / count case for avg).

Reply:
The reason I didn't have the interfaces/abstraction is to get rid of the references overhead that come with the interfaces. Enums/classes are lighter than the interfaces and since the Downsample iterators are super computational and memory intensive, tried to make them as simple as possible. Will put a TODO here and we shall profile and decide whether to generalize these Data structures.

Reply:
we chatted, TODO is fine to get started.